src/content/en/2021/javascript.md

```diff
@@ -1,20 +1,911 @@
+
+The usage was so frequent that [11.4%](../2020/javascript#how-do-we-load-our-javascript) of requests made on the mobile pages were seen with at least one script with `async` and `defer` attributes used together. The [root causes](https://twitter.com/rick_viscomi/status/1331735748060524551?s=20) were found and an action item was also taken down to [remove such usage going forward](https://twitter.com/Kraft/status/1336772912414601224?s=20).
+
+This year, we improved our measures of calculating these figures by looking at the rendered DOM, as opposed to the initial HTML. Based on the custom metric run on the rendered DOM for pages using these attributes, the variation is a lot as compared to the previous year. When we dived deeper to understand this gap, it was found that a lot of pages update these attributes dynamically after the pages has already been loaded. This is one reason why more results are found using the results from the custom metric.
+
+{{ figure_markup(
+  content="35.6%",
+  caption="Percent of pages using the `async` and `defer` attribute together.",
+  classes="big-number",
+  sheets_gid="2038121216",
+  sql_file="breakdown_of_pages_using_async_defer.sql"
+  )
+}}
+
+It is found that 35.6% of mobile pages use the `async` and `defer` attributes together. This was counted using a <a hreflang="en" href="https://github.com/HTTPArchive/legacy.httparchive.org/blob/master/custom_metrics/javascript.js">custom metric</a> that measures the usage of these attributes on the rendered DOM.
+
+A typical <a hreflang="en" href="https://www.tntsupercenter.com/">example website</a>, loads the below script:
+
+```js
+<!-- Piwik -->
+<script type="text/javascript">
+   (function() {
+      var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
+      g.type='text/javascript'; g.async=true; g.defer=true; g.src=u+'piwik.js'; s.parentNode.insertBefore(g,s);
+   })();
+</script>
+<!-- End Piwik Code -->
+```
+
+According to [Wikipedia](https://en.wikipedia.org/wiki/Matomo_(software)), Piwik (now known as Matomo):
+<blockquote>is a free and open source web analytics application developed by a team of international developers, that runs on a PHP/MySQL webserver. It tracks online visits to one or more websites and displays reports on these visits for analysis. As of June 2018, Matomo was used by over 1,455,000 websites, or 1.3% of all websites with known traffic analysis tools...</blockquote>
+
+This strongly suggests that marketing and analytics providers are a common source of these async/defer scripts being dynamically injected into the page.
+
+{# TODO Why does this strongly suggest this? Is it the fact that they are obviously dynamically inserted? Is it only marketing and analytics providers that prodominately using dynamically inserted script tags like this? #}
```

Comment:
@NishuGoel

Reply:
tweaked the sentence a bit to make it sound simpler